### PR TITLE
Add Hadolint Dockerfile Linter to CI

### LINKED ISSUE #5332: Add Dockerfile linter to CI
Suggested by @rbradford from https://github.com/cloud-hypervisor/cloud-hypervisor/pull/5324#issuecomment-1492897077

As a summary, to ensure our Dockerfile is properly formatted and valid, we want to include a Dockerfile linter to our CI, either as a new Github Action or a new Jenkins step. Ideally, we just need to run the linter if and only if our Dockerfile is modified.

### PR DESCRIPTION
**Pull request by rmstoi**
Fixes #5332

### CODE REVIEW

**Comment by rbradford** [comment]
Can you add a commit to this PR that adds a comment to the Dockerfile that says it is linted. This should also make the linter run.

**Review by likebreath — COMMENTED** [review]
Will the linter catch what you reported earlier #5324? Assume yes, please also include the fixes (in a separate commit). Thanks!

**Comment by rmstoi** [comment]
> Can you add a commit to this PR that adds a comment to the Dockerfile that says it is linted. This should also make the linter run.

done

**Comment by rmstoi** [comment]
> Will the linter catch what you reported earlier #5324? Assume yes, please also include the fixes (in a separate commit). Thanks!

Unfortunately Dockerfile linters I tried, hadolint and npm's dockerfile_lint, do not catch whitespace issues. I added previous commit a3a046c6b3901dd70b1727ea3a54d78a7bb9aa03 to untabify the dockerfile.

**Comment by likebreath** [comment]
@rmstoi Two good practices for WIP PRs: 1) keep the PR as a draft before it is ready for final review and merge; 2) add 'WIP' prefix to skip the CI pipeline when it is not needed.

About the PR itself, please ensure the linter is working properly. I see the following errors/warnings from the linter while it passed anyway:

```
Error: ./resources/Dockerfile:20 DL3008 warning: Pin versions in apt get install. Instead of `apt-get install <package>` use `apt-get install <package>=<version>`
Error: ./resources/Dockerfile:20 DL3015 info: Avoid additional packages by specifying `--no-install-recommends`
Error: ./resources/Dockerfile:60 DL3008 warning: Pin versions in apt get install. Instead of `apt-get install <package>` use `apt-get install <package>=<version>`
Error: ./resources/Dockerfile:60 DL3015 info: Avoid additional packages by specifying `--no-install-recommends`
Error: ./resources/Dockerfile:70 DL3008 warning: Pin versions in apt get install. Instead of `apt-get install <package>` use `apt-get install <package>=<version>`
Error: ./resources/Dockerfile:70 DL3015 info: Avoid additional packages by specifying `--no-install-recommends`
Error: ./resources/Dockerfile:90 SC2086 info: Double quote to prevent globbing and word splitting.
Error: ./resources/Dockerfile:99 DL4006 warning: Set the SHELL option -o pipefail before RUN with a pipe in it. If you are using /bin/sh in an alpine image or if your shell is symlinked to busybox then consider explicitly setting your SHELL to /bin/ash, or disable this check
Error: ./resources/Dockerfile:99 SC2086 info: Double quote to prevent globbing and word splitting.
Error: ./resources/Dockerfile:111 SC2086 info: Double quote to prevent globbing and word splitting.
Error: ./resources/Dockerfile:117 DL3003 warning: Use WORKDIR to switch to a directory
Error: ./resources/Dockerfile:117 SC2006 style: Use $(...) notation instead of legacy backticks `...`.
Error: ./resources/Dockerfile:1[34](https://github.com/cloud-hypervisor/cloud-hypervisor/actions/runs/4993664915/jobs/8943110066?pr=5433#step:4:35) DL3003 warning: Use WORKDIR to switch to a directory
Error: ./resources/Dockerfile:153 DL3047 info: Avoid use of wget without progress bar. Use `wget --progress=dot:giga <url>`. Or consider using `-q` or `-nv` (shorthands for `--quiet` or `--no-verbose`).
```

**Comment by rmstoi** [comment]
> About the PR itself, please ensure the linter is working properly. I see the following errors/warnings from the linter while it passed anyway:
> 

@likebreath
Althought they start with an "Error:" those are `warning` and `info` messages. These messages are printed but the CI does not fail on those, because of [`failure-threshold: error`](https://github.com/cloud-hypervisor/cloud-hypervisor/pull/5433/commits/bc6c5b16357942134edf6f769508b4a13e0df8a5#diff-93a1f172ec6e4110cd725d72be37e9f155cffa1b2470bcdaa99bb49cfa2201d1R26).
Hence, Dockerfile lint workflow [passes](https://github.com/cloud-hypervisor/cloud-hypervisor/actions/runs/5002636844)

There are few option if you think these messages should be disabled:
- disable verbose output of hadolint and these messages
- fix these info/warning issues in the Dockerfile
- do nothing, (CI passes even if these messages are printed)

Which would be more suitable option?

**Comment by likebreath** [comment]
> > About the PR itself, please ensure the linter is working properly. I see the following errors/warnings from the linter while it passed anyway:
> 
> @likebreath @rbradford Althought they start with an "Error:" those are `warning` and `info` messages. These messages are printed but the CI does not fail on those, because of [`failure-threshold: error`](https://github.com/cloud-hypervisor/cloud-hypervisor/pull/5433/commits/bc6c5b16357942134edf6f769508b4a13e0df8a5#diff-93a1f172ec6e4110cd725d72be37e9f155cffa1b2470bcdaa99bb49cfa2201d1R26). Hence, Dockerfile lint workflow [passes](https://github.com/cloud-hypervisor/cloud-hypervisor/actions/runs/5002636844)
> 
> There are few option if you think these messages should be disabled:
> 
> * disable verbose output of hadolint and these messages
> * fix these info/warning issues in the Dockerfile
> * do nothing, (CI passes even if these messages are printed)
> 
> Which would be more suitable option?

I think you should elevate the failure-threshold to fail with warnings and fix these linter warnings.

File: .github/workflows/hadolint.yml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        uses: actions/checkout@v3
+
+      - name: Lint Dockerfile
+        uses: hadolint/hadolint-action@master
+        with:
+          dockerfile: ./resources/Dockerfile
+          format: tty
+          no-fail: false
+          verbose: true
+          failure-threshold: error
```
[the remark this answers is not in the document]
Comment (rmstoi) [reply .github/workflows/hadolint.yml]: I've now tried to fix most of issues reported by hadolint, and ignore those that cannot be fixed. I've set the set failure-threshold to the default: info.

**Review by likebreath — COMMENTED** [review]
Looking much better now. Couple of comments:
1. Please re-organize the commits in the order of: add Hadolint linter (single commit), fix linter warnings (multiple commits), and ignore non-fixable linter warnings (multiple commits).
2. Can you please revisit the linter warnings being ignored now, and see if we can fix them. If not, please document why.

File: resources/Dockerfile [hunk 2]
```diff
@@ -18,54 +18,54 @@ ENV PATH="$PATH:$CARGO_HOME/bin"
...
-	&& apt-get clean \
-    && rm -rf /var/lib/apt/lists/*;	fi
+        apt-get update \
+        && apt-get -yq upgrade \
+        && DEBIAN_FRONTEND=noninteractive apt-get install -yq gcc-multilib gawk \
+        libtool expect gnutls-dev gnutls-bin libfuse-dev \
+        libjson-glib-dev libgmp-dev libtasn1-dev python3-twisted \
+        net-tools softhsm2 \
+        && apt-get clean \
+    && rm -rf /var/lib/apt/lists/*;     fi
```
Comment (likebreath) [inline resources/Dockerfile:73]: Redundant whitespace.

File: resources/Dockerfile [hunk 3]
```diff
@@ -84,7 +84,7 @@ RUN if [ "$TARGETARCH" = "arm64" ]; then \
         perl \
         texinfo \
         && apt-get clean \
-        && rm -rf /var/lib/apt/lists/*;	fi
+        && rm -rf /var/lib/apt/lists/*;         fi
```
Comment (likebreath) [inline resources/Dockerfile:93]: Redundant whitespace.

**Review by likebreath — COMMENTED** [review]
Looking good. Just started the CI pipeline for this PR.

Please update the container version from our scripts. Also two nits below.

**Comment by likebreath** [comment]
Just realized that you will need to change the container version tag from the scripts, so that the CI pipeline can re-build the container image with the updated Dockerfile and run tests with it.

So please update the container version from the scripts and push again to trigger the CI pipeline.

**Comment by rmstoi** [comment]
> Just realized that you will need to change the container version tag from the scripts, so that the CI pipeline can re-build the container image with the updated Dockerfile and run tests with it.
> 
> So please update the container version from the scripts and push again to trigger the CI pipeline.

I've updated `CTR_IMAGE_VERSION="20230526-0"`  in `scripts/dev_cli.sh`

**Comment by rmstoi** [comment]
@likebreath some tests fail, could it be because new docker image is not used?

**Comment by rbradford** [comment]
@rmstoi If you check the log you can see it building the container: https://cloud-hypervisor-jenkins.westus.cloudapp.azure.com/blue/rest/organizations/jenkins/pipelines/cloud-hypervisor/branches/PR-5433/runs/21/nodes/47/steps/135/log/?start=0

**Comment by likebreath** [comment]
@rmstoi Have you tried building the container image with your updated Dockerfile locally? Both the aarch64 worker and the x86 worker are failing due to failures of building container image.

**Comment by rmstoi** [comment]
@rbradford  @likebreath  I do test container build with docker before push, and docker workflows also pass. Not sure if that was an issue with Jenkins, cause it looked like an old issue I fixed in this branch a while ago. After rebase and  push container build started to work but the test_ovs_dpdk  fails.  That seems like a regression, likely caused by my changes.

**Comment by likebreath** [comment]
@rmstoi Right, the ovs-dpdk test replies heavily on the docker image, and likely your changes make a difference here causing the test failure. Please run the dpdk tests locally and see if you can reproduce and get it fixed.

Converting the PR to a draft.

**Comment by rbradford** [comment]
The linter should only enabled after all the fixes that it needs are made.

**Comment by rmstoi** [comment]
> Please reorganize the PR commits in the following way:
> 
>     1. Squash all commits related to add required packages to the commit "Dockerfile: fix hadolint error, do not install recommended packages";

done

> 
>     2. Squash all commits of "ignore hadolint warning" into one commit;
> 

done

> 
> Also, I am assuming the reason to add "--no-install-recommends" is reducing the size of container image. If so, what's the image size reduction we got from this change?

My guess is around 20M (`docker images`). Could be easier see when docker image is in registry.

**Comment by likebreath** [comment]
> My guess is around 20M (docker images). Could be easier see when docker image is in registry.

Well, you can easily measure it by building the container image locally. 

It actually should have been done earlier to help you decide whether it is worth adding "--no-install-recommends" over simply ignore it as what's done for many other warnings, particularly given this is the major cause of lots of test failures you have been debugging on.

> The linter should only enabled after all the fixes that it needs are made.

Please address Rob's comments above. It is normally expected to have all pending comments addressed before a new push, which can avoid unnecessary runs of our CI pipeline.

**Comment by rmstoi** [comment]
> The linter should only enabled after all the fixes that it needs are made.

Should I disable the hadolint for now?
What do you mean by "all the fixes that it needs are made"?

**Comment by rbradford** [comment]
> > The linter should only enabled after all the fixes that it needs are made.
> 
> Should I disable the hadolint for now? What do you mean by "all the fixes that it needs are made"?

I'm referring to the commit order.

**Comment by rbradford** [comment]
It's possible that test_ovs_dpdk is still failing on aarch64 - since it could be a flake I have retriggered.

**Comment by likebreath** [comment]
> Nevertheless it is a good idea to use "--no-install-recommends" to make sure that the cloud hypervisor docker image has only needed apt packages.

If this is a really important best practice to follow and you have spent lots of time on it, I would expect you have addressed all related issues in the PR. But I still see one annotation being made to ignore this linter warning:
```
# DL3015 ignored cause not installing openvswitch-switch-dpdk recommended packages breaks ovs_dpdk test
```

Anyway, not a big issue to me. I will let you decide if you follow through this issue. Landing the PR now. Thank you for the contribution.

**Pull request merged**